k8s-test-runner/k8s_client/client.go

```diff
@@ -0,0 +1,254 @@
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	K8sStatePollInterval = 3 * time.Second
+)
+
+// High level k8s client
+type Client struct {
+	ClientSet  *kubernetes.Clientset
+	RESTConfig *rest.Config
+}
+
+// GetLocalK8sDeps get local k8s context config
+func GetLocalK8sDeps() (*kubernetes.Clientset, *rest.Config, error) {
+	loadingRules := clientcmd.NewDefaultClientConfigLoadingRules()
+	kubeConfig := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(loadingRules, &clientcmd.ConfigOverrides{})
+	k8sConfig, err := kubeConfig.ClientConfig()
+	if err != nil {
+		return nil, nil, err
+	}
+	k8sClient, err := kubernetes.NewForConfig(k8sConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+	return k8sClient, k8sConfig, nil
+}
+
+// NewK8sClient creates a new k8s client with a REST config
+func NewClient() *Client {
+	cs, cfg, err := GetLocalK8sDeps()
+	if err != nil {
+		log.Fatal().Err(err).Send()
+	}
+	return &Client{
+		ClientSet:  cs,
+		RESTConfig: cfg,
+	}
+}
+
+// k8sOperation defines the function type for Kubernetes operations that need retries.
+type k8sOperation func() error
+
+// retryK8sCall attempts the provided Kubernetes operation with retries.
+func retryK8sCall(operation k8sOperation, maxRetries int) error {
+	retryPolicy := wait.Backoff{
+		Steps:    maxRetries,       // Max number of retries.
+		Duration: 10 * time.Second, // Initial delay before the first retry.
+		Factor:   2.0,              // Multiplier factor for subsequent delays.
+		Jitter:   0.1,              // Random jitter for the delay.
+	}
+
+	var lastError error
+	err := wait.ExponentialBackoff(retryPolicy, func() (bool, error) {
+		lastError = operation()
+		if lastError != nil {
+			log.Warn().Err(lastError).Msg("Error encountered during K8s call, will retry")
+			return false, nil
+		}
+		return true, nil
+	})
+
+	if err != nil {
+		return fmt.Errorf("after %d attempts, last error: %s", maxRetries, lastError)
+	}
+	return nil
+}
+
+func (m *Client) ListPods(ctx context.Context, namespace, syncLabel string) (*v1.PodList, error) {
+	var pods *v1.PodList
+	maxRetries := 5 // Maximum number of retries
+	timeout := int64(30)
+	labelSelector := syncSelector(syncLabel)
+	operation := func() error {
+		var err error
+		pods, err = m.ClientSet.CoreV1().Pods(namespace).List(ctx, metaV1.ListOptions{
+			LabelSelector:  labelSelector,
+			TimeoutSeconds: &timeout,
+		})
+		return err
+	}
+
+	err := retryK8sCall(operation, maxRetries)
+	if err != nil {
+		// Wrap and return any error encountered during the retry operation
+		return nil, fmt.Errorf(`failed to call CoreV1().Pods().List(), namespace: %s, labelSelector: %s, timeout: %d: %w`, namespace, labelSelector, timeout, err)
+	}
+
+	// At this point, `pods` should be populated successfully
+	return pods, nil
+}
+
+func (m *Client) ListJobs(ctx context.Context, namespace, syncLabel string) (*batchV1.JobList, error) {
+	var jobs *batchV1.JobList
+	maxRetries := 5 // Maximum number of retries
```

Comment:
Hoist it to the package level const, both retries and timeout. Or even add it to `config.go`, all k8s methods

Reply:
done. Also updated waiting logic to work with CI